xdp-dump/xdpdump.c

```diff
@@ -947,7 +947,14 @@ static int find_target(struct dumpopt *cfg, struct xdp_multiprog *mp,
 	char                    *prog_safe_ptr;
 	char                    *program_names = cfg->program_names;
 
-	prog = xdp_multiprog__main_prog(mp);
+	struct xdp_program *first_prog = xdp_multiprog__next_prog(NULL, mp);
+	struct xdp_program *second_prog = xdp_multiprog__next_prog(first_prog, mp);
+	bool freplaced_dispatcher = first_prog != NULL && second_prog == NULL;
+	if (freplaced_dispatcher) {
+		prog = first_prog;
+	} else {
+		prog = xdp_multiprog__main_prog(mp);
+	}
```

Comment:
Hmm, I wonder if  instead of detecting this case, we could just make xdpdump always attach to the first component program if there is only one? Shouldn't make any difference for the dumped packets, and it makes things work regardless of whether the freplace target was changed or not (so it's backwards compatible).

@chaudron any objections to that?

Reply:
Trying to understand why we need this change at all? If there is a dispatcher XDP dump should attach to the dispatcher, if there is none it will be the first program. Or do I miss something?

Reply:
See the previous commit: the idea is to change the libxdp logic so that if there's only a single program, we'll just freplace the main function of the dispatcher instead of the prog0 function that it calls. This saves the overhead of calling multiple functions, which hopefully makes the performance overhead of the dispatcher go away for a single program; but it means that the main dispatcher function never gets called at all, so attaching to it from xdpdump no longer works...

Reply:
Doesn't this change mess up the entire API? Because now we think two programs are executed/loaded, but we only have one? It might also confuse users, for example for a user displaying the iface_print_status() output makes him feel like we still use the dispatcher?

Can we not have xdp_multiprog__main_prog() return the first (and only) attached program in this case? It will solve the problem for xdpdump, and would not require any changes.

Reply:
Well, my feeling is that it's sort of an under-the-hood detail that I'm not sure if we should be exposing at all? I mean from a semantic PoV, there's still a dispatcher that dispatches to a single program, it's just that we changed the details of how the call is made to avoid some overhead?

So from an API PoV, you still have main_prog() is the dispatcher which is the ID that is attached to the interface; and the actual user progs are returned by __next_prog() and those are what you should be looking at to find what is actually processing the packets. So really, making assumptions about the main program is a bit of a niche application that we're only doing in xdpdump, so it's easier to just fix that?

Changing main_prog() to return the first program has the problem that then there would be no way to get at the actual dispatcher program; which is still the program ID attached to the interface, even if it is never executed. So if someone wanted to, say, check if the main prog was attached to an interface, they'd need a reference to the dispatcher.

Also, there's an additional problem in that we can't actually detect if this optimisation has been applied: the kernel has no interface to discover the target of an attached freplace program. I'm going to submit a patch to fix that, but for older kernels that obviously doesn't help. So if we are going to track that in a backwards-compatible way, we'll need to keep that information somewhere like bpffs; or we could bump the dispatcher version and deal with it that way (but thus losing forward-compatibility for old versions of the library)...

Reply:
Let me take another look tomorrow, the first commit has no comments, and I'm having a problem trying to grasp it (guess I need to sleep or get coffee...)

Reply:
Heh, yeah, I see what you mean, it's missing a bit of context. @shoracek I guess you should take that as feedback: Please write a commit message for the first commit explaining what it's doing and why.

@chaudron see #103 for a bit of that missing context :)

Reply:
@shoracek I looked at your change, and I think it can be more clean doing Toke's suggestion. Here is a diff you could apply that will do all that is needed from the tcpdump side:

```
diff --git a/xdp-dump/xdpdump.c b/xdp-dump/xdpdump.c
index 69a703d..c398b19 100644
--- a/xdp-dump/xdpdump.c
+++ b/xdp-dump/xdpdump.c
@@ -947,17 +947,21 @@ static int find_target(struct dumpopt *cfg, struct xdp_multiprog *mp,
        char                    *prog_safe_ptr;
        char                    *program_names = cfg->program_names;
 
-       struct xdp_program *first_prog = xdp_multiprog__next_prog(NULL, mp);
-       struct xdp_program *second_prog = xdp_multiprog__next_prog(first_prog, mp);
-       bool freplaced_dispatcher = first_prog != NULL && second_prog == NULL;
-       if (freplaced_dispatcher) {
-               prog = first_prog;
-       } else {
-               prog = xdp_multiprog__main_prog(mp);
-       }
+       prog = xdp_multiprog__main_prog(mp);
 
        /* First take care of the default case, i.e. no function supplied */
        if (!program_names) {
+               struct xdp_program *first_prog;
+
+               /* The libxdp code optimization where it skips the dispatcher
+                * if only one program is loaded. If this is the case, we need
+                * to attach to the actual first program, not the dispatcher.
+                */
+               first_prog = xdp_multiprog__next_prog(NULL, mp);
+               if (first_prog != NULL &&
+                   xdp_multiprog__next_prog(first_prog, mp) == NULL)
+                       prog = first_prog;
+
                matches = find_func_matches(xdp_program__btf(prog),
                                            xdp_program__name(prog),
                                            &func, false, -1, false);
@@ -1006,23 +1010,21 @@ static int find_target(struct dumpopt *cfg, struct xdp_multiprog *mp,
                progs.progs[progs.nr_of_progs].func = NULL;
        progs.nr_of_progs++;
 
-       if (!freplaced_dispatcher) {
-               for (p = xdp_multiprog__next_prog(NULL, mp);
-                    p;
-                    p = xdp_multiprog__next_prog(p, mp)) {
-
-                       progs.progs[progs.nr_of_progs].prog = p;
-                       matches = find_func_matches(xdp_program__btf(p),
-                                                   xdp_program__name(p),
-                                                   &progs.progs[progs.nr_of_progs].func,
-                                                   false, -1, false);
-                       if (matches != 1)
-                               progs.progs[progs.nr_of_progs].func = NULL;
-                       progs.nr_of_progs++;
-
-                       if (progs.nr_of_progs >= MAX_LOADED_XDP_PROGRAMS)
-                               break;
-               }
+       for (p = xdp_multiprog__next_prog(NULL, mp);
+            p;
+            p = xdp_multiprog__next_prog(p, mp)) {
+
+               progs.progs[progs.nr_of_progs].prog = p;
+               matches = find_func_matches(xdp_program__btf(p),
+                                           xdp_program__name(p),
+                                           &progs.progs[progs.nr_of_progs].func,
+                                           false, -1, false);
+               if (matches != 1)
+                       progs.progs[progs.nr_of_progs].func = NULL;
+               progs.nr_of_progs++;
+
+               if (progs.nr_of_progs >= MAX_LOADED_XDP_PROGRAMS)
+                       break;
        }
 
        /* If "all" option is specified create temp program names */
```

The changes in the test cases are fine, as there are still some cases having multiple programs loaded I think no new test case is needed for this new behavior.

Reply:
To maybe make this cleaner:

```
+ first_prog = xdp_multiprog__next_prog(NULL, mp);
+ if (first_prog != NULL && + xdp_multiprog__next_prog(first_prog, mp) == NULL) 
+     prog = first_prog; 
```

We could add an API to return the number of programs loaded? i.e. xdp_multiprog__nr_loaded_program().

So it will become:
```
+ if (dp_multiprog__nr_loaded_program(mp) == 1) 
+     prog = first_prog; 
```

Reply:
Yeah, fine with adding that to the API... bikeshedding the name a bit, maybe `xdp_multiprog__program_count()` ?

Reply:
Ok, so the request to @shoracek would be to add the xdp_multiprog__program_count() API to the first commit and take my chances with using the xdp_multiprog__program_count() for the second commit. And add some more details to the commit message for both. @tohojo did I miss anything?

Reply:
No, that sounds about right! AFAICT :)

Reply:
xdpdump changes look good to me!